scripts/safe_template_linter.py

```diff
@@ -1406,6 +1419,375 @@ def _is_jquery_insert_caller_safe(self, caller_line_start):
+            self.generic_visit(node)
+
+
+class HtmlStringVisitor(BaseVisitor):
+    """
+    Checks for strings that contain HTML. Assumes any string with < or > is
+    considered potential HTML.
+
+    To be used only with strings in context of format or concat.
+
+    """
+    def __init__(self, file_contents, results, skip_wrapped_html=False):
+        """
+        Init function.
+
+        Arguments:
+            file_contents: The contents of the Python file.
+            results: A file results objects to which violations will be added.
+            skip_wrapped_html: True if visitor should skip strings wrapped with
+                HTML() or Text(), and False otherwise.
+        """
+        super(HtmlStringVisitor, self).__init__(file_contents, results)
+        self.skip_wrapped_html = skip_wrapped_html
+        self.unsafe_html_string_nodes = []
+        self.over_escaped_entity_string_nodes = []
+        self.has_text_or_html_call = False
+
+    def visit_Str(self, node):
+        """
+        When strings are visited, checks if it contains HTML.
+
+        Arguments:
+            node: An AST node.
+        """
+        # Skips '<' (and '>') in regex named groups. For example, "(?P<group>)".
+        if re.search('[(][?]P<', node.s) is None and re.search('[<>]', node.s) is not None:
+            self.unsafe_html_string_nodes.append(node)
+        if re.search(r"&[#]?[a-zA-Z0-9]+;", node.s):
+            self.over_escaped_entity_string_nodes.append(node)
+
+    def visit_Call(self, node):
+        """
+        Skips processing of string contained inside HTML() and Text() calls when
+        skip_wrapped_html is True.
+
+        Arguments:
+            node: An AST node.
+
+        """
+        is_html_or_text_call = isinstance(node.func, ast.Name) and node.func.id in ['HTML', 'Text']
+        if self.skip_wrapped_html and is_html_or_text_call:
+            self.has_text_or_html_call = True
+        else:
+            self.generic_visit(node)
+
+
+class CloseBeforeFormatVisitor(BaseVisitor):
+    """
+    HTML() and Text() should not have a format() anywhere in its argument. This
+    visitor can used to ensure there was no call to format().
+
+    This visitor is meant to be called on HTML() and Text() ast.Call nodes.
+
+    """
+    def __init__(self, file_contents, results):
+        """
+        Init function.
+
+        Arguments:
+            file_contents: The contents of the Python file.
+            results: A file results objects to which violations will be added.
+
+        """
+        super(CloseBeforeFormatVisitor, self).__init__(file_contents, results)
+        self.contains_format_call = False
+
+    def visit_Attribute(self, node):
+        """
+        Simple check for format calls (attribute).
+
+        Arguments:
+            node: An AST node.
+
+        """
+        # Attribute(expr value, identifier attr, expr_context ctx)
+        if node.attr == 'format':
+            self.contains_format_call = True
```

Comment:
@nedbat: I think this will just have to be ok.  
1. I imagine this isn't really causing an issue.
2. Unless you can help me see otherwise, format is an `ast.Attribute`, for which it is really easy to determine the left hand side (i.e. what is this an attribute of), but very difficult to get the right-hand side (i.e. what arguments are we passing to format()).

At this point, I plan on leaving this as-is.  Sound ok?